src/sparse/impl/KokkosSparse_twostage_gauss_seidel_impl.hpp

```diff
@@ -0,0 +1,919 @@
+        }
+#ifdef KOKKOSSPARSE_IMPL_TIME_TWOSTAGE_GS
+        Kokkos::fence();
+        tic = timer.seconds ();
+        std::cout << "TWO-STAGE GS::SYMBOLIC::INSERT TIME    : " << tic << std::endl;
+        timer.reset();
+#endif
+
+        // construct CrsMat with them
+        graph_t graphL (column_viewL, rowmap_viewL);
+        graph_t graphU (column_viewU, rowmap_viewU);
+        crsmat_t crsmatL ("L", num_rows, values_viewL, graphL);
+        crsmat_t crsmatU ("U", num_rows, values_viewU, graphU);
+
+        // store them in handle
+        gsHandle->setL (crsmatL);
+        gsHandle->setU (crsmatU);
+        gsHandle->setD (viewD);
+        if (!(gsHandle->isTwoStage ())) {
+          // create SpTRSV handles for classical GS
+          using namespace KokkosSparse::Experimental;
+          // TODO: figure out how to use CuSPARSE
+          #if 1 // !defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) || !defined(KOKKOS_ENABLE_CUDA)
+          using namespace KokkosSparse::Experimental;
+          sptrsv_symbolic (handle->get_gs_sptrsvL_handle(), rowmap_viewL, crsmatL.graph.entries);
+          sptrsv_symbolic (handle->get_gs_sptrsvU_handle(), rowmap_viewL, crsmatU.graph.entries);
+          #endif
+        }
+      }
+
+
+      /**
+       * Numerical setup
+       */
+      void initialize_numeric ()
+      {
+#ifdef KOKKOSSPARSE_IMPL_TIME_TWOSTAGE_GS
+        double tic;
+        Kokkos::Impl::Timer timer;
+        Kokkos::fence();
+        timer.reset();
+#endif
+        using GS_Functor_t = TwostageGaussSeidel_functor<const_row_map_view_t, entries_view_t, values_view_t>;
+
+        auto *gsHandle = get_gs_handle();
+        bool two_stage = gsHandle->isTwoStage ();
+
+        // load local D from handle
+        auto viewD = gsHandle->getD ();
+
+        // load local L from handle
+        auto crsmatL = gsHandle->getL ();
+        auto values_viewL = crsmatL.values;
+        auto rowmap_viewL = crsmatL.graph.row_map;
+
+        // load local U from handle
+        auto crsmatU = gsHandle->getU ();
+        auto values_viewU = crsmatU.values;
+        auto rowmap_viewU = crsmatU.graph.row_map;
+
+        // extract local L, D & U matrices
+        using range_policy = Kokkos::RangePolicy <Tag_valuesLU, execution_space>;
+        Kokkos::parallel_for ("valueLU", range_policy (0, num_rows),
+                              GS_Functor_t (two_stage, num_rows, rowmap_view, column_view, values_view, 
+                                                                 rowmap_viewL, values_viewL, viewD,
+                                                                 rowmap_viewU, values_viewU));
+#ifdef KOKKOSSPARSE_IMPL_TIME_TWOSTAGE_GS
+        Kokkos::fence();
+        tic = timer.seconds ();
+        std::cout << std::endl << "TWO-STAGE GS::NUMERIC::INSERT LU TIME : " << tic << std::endl;
+        timer.reset();
+#endif
+
+        if (!(gsHandle->isTwoStage ())) {
+          // TODO: figure out how to use CuSPARSE
```

Comment:
@iyamazaki it looks like there is already an interface to cusparse in `KokkosSparse_sptrsv.hpp` is there something preventing you from using it?

Reply:
Apparently you can do something like this:
```
sptrsv_handle.set_algorithm(SPTRSVAlgorithm::SPTRSV_CUSPARSE);
```